### ISSUE TITLE
Jobs that are redirected to another server by PBSPro are considered as "Failed" 

### ISSUE THREAD

**Issue by IdoBar** [nextflow-io/nextflow#1840]
## Bug report 

This issue seems similar to #412 . 
When running `nf-core/eager` on the QLD HPC system ([QRIS Awoonga](https://www.qriscloud.org.au/support/qriscloud-documentation/92-awoonga-user-guide), running `PBSpro`), jobs that get redirected to another queue (on another server) are being referred to as "failed" with this error message `terminated for an unknown reason -- Likely it has been terminated by the external system`, even tough they complete successfully with an exit status 0 (which nextflow is unable to get).

### Expected behavior and actual behavior

The workflow should keep tracing the jobs even if they're automatically redirected to another queue/server.
I'm not sure how nextflow is tracking job progress, but my guess is that it has to do with the job status being changed to `M` instead of `R` or that nextflow is using the full `job_id.server` to track the job progress rather that using just `job_id` (see output of `qstat` below)

This is the `qstat` summary of a job that had "failed" according to nextflow:
```
Job Id: 510838.awonmgr2
    Job_Name = nf-genotyping_f
    Job_Owner = ibar@awoonga1.local
    job_state = M
    queue = Short@flashmgr2
    server = awonmgr2
    Account_Name = qris-gu
    Checkpoint = u
    ctime = Fri Dec 25 09:28:22 2020
    Error_Path = awoonga1.local:/gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708/nf-genotyping_f.e510838
    Hold_Types = n
    Join_Path = oe
    Keep_Files = n
    Mail_Points = a
    mtime = Fri Dec 25 09:29:59 2020
    Output_Path = awoonga1.local:/gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708/.command.log
    Priority = 0
    qtime = Fri Dec 25 09:28:22 2020
    Rerunable = True
    Resource_List.interact1 = 0
    Resource_List.mem = 4096mb
    Resource_List.ncpus = 2
    Resource_List.nodect = 1
    Resource_List.place = free
    Resource_List.select = 1:ncpus=2:mem=4096mb
    Resource_List.walltime = 04:00:00
    substate = 92
    Variable_List = PBS_O_SYSTEM=Linux,PBS_O_SHELL=/bin/bash,PBS_O_HOME=/home/ibar,PBS_O_LOGNAME=ibar,PBS_O_WORKDIR=/gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708,PBS_O_LANG=en_AU.UTF-8,PBS_O_PATH=/home/ibar/.pyenv/bin:/sw/Containers/singularity/3.5.0/bin:/sw/Containers/singularity/images:/home/ibar/.pyenv/bin:/home/ibar/bin:/home/ibar/.pyenv/versions/miniconda-latest/bin:/home/ibar/.pyenv/versions/miniconda-latest/condabin:/home/ibar/.pyenv/bin:/home/ibar/bin:/opt/gnu/gcc/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/ganglia/bin:/opt/ganglia/sbin:/usr/java/latest/bin:/opt/pbs/bin:/opt/rocks/bin:/opt/rocks/sbin:/home/ibar/etc/tools/Annotation/BLAT:/opt/pbs/bin:/home/ibar/etc/tools/Annotation/BLAT,PBS_O_MAIL=/var/spool/mail/ibar,PBS_O_QUEUE=workq,PBS_O_HOST=awoonga1.local
    comment = Job finished at "flashmgr2"
    etime = Fri Dec 25 09:28:22 2020
    eligible_time = 35:46:34
    Submit_arguments = -N nf-genotyping_f .command.run
    history_timestamp = 1608852599
    project = _pbs_project_default
```

### Steps to reproduce the problem

Run a nextflow workflow on Awoonga QRIS HPC.
Example of my command:
```
nextflow run nf-core/eager \
		 -r 2.2.2 \
		 -params-file Dingo_aDNA_NF1.CanFam3.1.bt2local.freebayes.json \
		 -c /home/ibar/.nextflow/awoonga.config
``` 
### Program output 

These are the relevant messages in the log file:
```
Dec.-25 09:37:12.451 [Task monitor] DEBUG nextflow.executor.GridTaskHandler - Failed to get exit status for process TaskHandler[jobId: 510838.awonmgr2; id: 122; name: genotyping_freebayes (D10); status: RUNNING; exit: -; error: -; workDir: /gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708 started: 1608852607445; exited: -; ] -- exitStatusReadTimeoutMillis: 270000; delta: 270020
Current queue status:
>   job: 510842.awonmgr2: PENDING
>   job: 510843.awonmgr2: PENDING
>   job: 510844.awonmgr2: PENDING
>   job: 510845.awonmgr2: PENDING
>   job: 510846.awonmgr2: PENDING
>   job: 510847.awonmgr2: PENDING
>   job: 510848.awonmgr2: PENDING
>   job: 510849.awonmgr2: PENDING
>   job: 510850.awonmgr2: PENDING
>   job: 510851.awonmgr2: PENDING
>   job: 510852.awonmgr2: PENDING
>   job: 510853.awonmgr2: PENDING
>   job: 510854.awonmgr2: PENDING
>   job: 510855.awonmgr2: PENDING

Content of workDir: /gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708
null
Dec.-25 09:37:12.452 [Task monitor] DEBUG n.processor.TaskPollingMonitor - Task completed > TaskHandler[jobId: 510838.awonmgr2; id: 122; name: genotyping_freebayes (D10); status: COMPLETED; exit: -; error: -; workDir: /gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708 started: 1608852607445; exited: -; ]
Dec.-25 09:37:12.468 [Task monitor] DEBUG nextflow.processor.TaskRun - Unable to dump output of process 'genotyping_freebayes (D10)' -- Cause: java.nio.file.NoSuchFileException: /gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708/.command.out
Dec.-25 09:37:12.482 [Task monitor] DEBUG nextflow.processor.TaskRun - Unable to dump error of process 'genotyping_freebayes (D10)' -- Cause: java.nio.file.NoSuchFileException: /gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708/.command.err
Dec.-25 09:37:12.483 [Task monitor] DEBUG nextflow.processor.TaskRun - Unable to dump error of process 'genotyping_freebayes (D10)' -- Cause: java.nio.file.NoSuchFileException: /gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708/.command.log
Dec.-25 09:37:12.486 [Task monitor] ERROR nextflow.processor.TaskProcessor - Error executing process > 'genotyping_freebayes (D10)'
Caused by:
  Process `genotyping_freebayes (D10)` terminated for an unknown reason -- Likely it has been terminated by the external system
Command executed:
  freebayes -f CanFam3.1.fasta -p 2 -C 5 -g 0 D10.trimmed.bam > D10.freebayes.vcf
  pigz -p 2 D10.freebayes.vcf
Command exit status:
  -
Command output:
  (empty)

Work dir:
  /gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708

Tip: view the complete command output by changing to the process work dir and entering the command `cat .command.out`
Dec.-25 09:37:12.493 [Task monitor] INFO  nextflow.Session - Execution cancelled -- Finishing pending tasks before exit
Dec.-25 09:37:12.551 [Task monitor] DEBUG nextflow.processor.TaskRun - Unable to dump error of process 'genotyping_freebayes (D10)' -- Cause: java.nio.file.NoSuchFileException: /gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708/.command.err
Dec.-25 09:37:12.552 [Task monitor] DEBUG nextflow.processor.TaskRun - Unable to dump output of process 'genotyping_freebayes (D10)' -- Cause: java.nio.file.NoSuchFileException: /gpfs1/scratch/30days/ibar/data/Dingo/Dingo_aDNA_NF13_process_24_12_2020/work/83/4cb7e477c5ad7a5236ba0d6d314708/.command.out
```

And the content of `.command.log`:
```
########################### Execution Started #############################
JobId:510838.awonmgr2
UserName:ibar
GroupName:qris-gu
ExecutionHost:fl118
###############################################################################
nxf-scratch-dir fl118.local:/nvme/pbs/tmpdir/pbs.510838.awonmgr2/nxf.urortU9s5E
########################### Job Execution History #############################
JobId:510838.awonmgr2
UserName:ibar
GroupName:qris-gu
JobName:nf-genotyping_f
SessionId:11147
ResourcesRequested:mem=4096mb,ncpus=2,place=free,walltime=04:00:00
ResourcesUsed:cpupercent=89,cput=00:12:28,mem=158040kb,ncpus=2,vmem=1129996kb,walltime=00:12:25
QueueUsed:Short
AccountString:qris-gu
ExitStatus:0
###############################################################################
```
### Environment 

* Nextflow version: v20.10.0 
* Java version: OpenJDK Runtime Environment Zulu11.2+3 (build 11.0.1+13-LTS)
* Operating system: CentOS Linux release 7.4.1708
* Bash version: GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)

### Additional context

This happens when jobs are being redirected from the main server (`awonmgr2`) to the `flashmgr2` server.
Happy to provide additional information if needed.

Thanks, Ido

**Comment by pditommaso** [comment]
NF uses the `qstat -f <queue name>` to retrieve status for the job(s). If the job is moved to another queue, it won't be found any more. 

Does `qstat` allow the specification of more than one queue name?

**Comment by chewgl** [comment]
I'm currently running into the same issue where my jobs are submitted to an external queue, and then rerouted to an internal queue, and then cannot be found with qstat -f <queue name>. Could qstat -f be run without the queue name instead as a way to work-around this issue?

**Comment by chewgl** [comment]
Another work-around (without code changes) might be to supply the queue to be submitted in the config file, under clusterOptions = '-q <submit-queue-name>' such that there is no queue name data stored, and thus qstat -f runs without the queue name. It seems to work for me; @IdoBar: perhaps try this?

**Comment by IdoBar** [comment]
Thanks @chewgl, I'll give it a try.
@pditommaso, is it possible to query the queue by the user (`-u username`) or more specifically using just the job_id (while removing the queue name)?
I'll have a look at `nextflow/src/main/groovy/nextflow/executor/PbsProExecutor.groovy` to see if I have any ideas to experiment with.

Thanks, Ido

**Comment by pditommaso** [comment]
> Does qstat allow the specification of more than one queue name?

**Comment by IdoBar** [comment]
Yes @pditommaso, seems like you can provide multiple queue names at the end of the command: `qstat -f queue@server queue@server2`.
One thing that's worth mentioning, is that in the HPC that I work, `qstat` is modified from its default behavior and it shows only my jobs, even without using the `-u username` option. In addition, it shows the status of all servers when used with the `-q` flag.
Example output of `qstat -q`:
```
server: awonmgr2.storage

Queue            Memory CPU Time Walltime Node   Run   Que   Lm  State
---------------- ------ -------- -------- ---- ----- ----- ----  -----
DeadEnd            --      --       --     --      0     0   --   E S
workq              --      --       --     --      0     0   --   E R
CVL                62gb    --    168:00:0    1     2     0   --   E R
Interact          250gb    --    48:00:00    1     3     0   --   E R
GPU                64gb    --    48:00:00    1     0     0   --   D R
Long              250gb    --    336:00:0    1    32     2   --   E R
Single            250gb    --    168:00:0    1    90    33   --   E R
Short             250gb    --    24:00:00    1    50   101   --   E R
test               --      --       --     --      0     0   --   D R
Special           120gb    --    01:00:00    1     0     1   --   E R
                                               ----- -----
                                                 177   137

server: flashmgr2.ipoib

Queue            Memory CPU Time Walltime Node   Run   Que   Lm  State
---------------- ------ -------- -------- ---- ----- ----- ----  -----
DeadEnd            --      --       --     --      0     0   --   E S
Interact          500gb    --    24:00:00    1     4     0   --   E R
BeeGFS2            --      --       --     --      0     0   --   E R
test               --      --       --     --      0     0   --   D R
workq              --      --       --     --      0     0   --   E R
Long              500gb    --    336:00:0    1     2     1   --   E R
Single            500gb    --    168:00:0    1    22    11   --   E R
Multiple           --      --    168:00:0   52     0     0   --   E R
Special           957gb    --    168:00:0    1     0     0   --   E R
ShortMemory       957gb    --    24:00:00    1     0     0   --   E R
Remote             --      --       --     --      0     0   --   D S
Short             500gb    --    24:00:00  --     51    18   --   E R
BigMemory           4tb    --    336:00:0    1     3     2   --   E R
BigSMP             --      --       --       1     0     1   --   E R
BeeGFS             --      --    672:00:0   10     1     0   --   E R
                                               ----- -----
                                                  83    33

```

**Comment by IdoBar** [comment]
@chewgl,  

> 
> 
> Another work-around (without code changes) might be to supply the queue to be submitted in the config file, under clusterOptions = '-q ' such that there is no queue name data stored, and thus qstat -f runs without the queue name. It seems to work for me; @IdoBar: perhaps try this?

This doesn't work because it changes the behavior of `qsub` as well, which will fail if `-q` is provided without an argument.

**Comment by chewgl** [comment]
@IdoBar : I meant to specify the queue name in
`clusterOptions = '-q [submit-queue-name]'` in a config file, rather than as part of `queue = [submit-queue-name]`; these arguments in clusterOptions are still passed to qsub, so it should work, but doesn't store it as a variable under nextflow.

I can confirm that it works on my HPC running PBSpro, but I realize that different HPCs might be configured slightly differently.

I agree that being able to specify a user name might be helpful.

**Comment by pditommaso** [comment]
Querying the status by job id would trigger the execution of too many processes for large workloads. Not a good option. 

I think there could be two possible solution

1. modifying the `qstat` command executed by NF removing the queue information and adding the username (requires changing the NF code) 
2. provided that both `qstat` and `qsub` allow the specification of multiple queue names, provide such names in the nextflow `process.queue` option. This *may* work with the current version, tho it should be verified.

**Comment by IdoBar** [comment]
> 
> 
> Querying the status by job id would trigger the execution of too many processes for large workloads. Not a good option.
> 
> I think there could be two possible solution
> 
>     1. modifying the `qstat` command executed by NF removing the queue information and adding the username (requires changing the NF code)
> 
>     2. provided that both `qstat` and `qsub` allow the specification of multiple queue names, provide such names in the nextflow `process.queue` option. This _may_ work with the current version, tho it should be verified.

I used the following code from [PbsProExecutor.groovy](https://github.com/nextflow-io/nextflow/blob/master/modules/nextflow/src/main/groovy/nextflow/executor/PbsProExecutor.groovy) to parse the job status from the queue:
```
@Override
    protected List<String> queueStatusCommand(Object queue) {
        String cmd = 'qstat -f'
        if( queue ) cmd += ' ' + queue
        return ['bash','-c', "set -o pipefail; $cmd | { egrep '(Job Id:|job_state =)' || true; }".toString()]
    }
```
In my HPC system, the job that was moved will not appear with `qstat -f | egrep '(Job Id:|job_state =)' `, unless I specify the server it was moved to (flashmgr2) with `qstat -f @flashmgr2 | egrep '(Job Id:|job_state =)' ` and only then I can see the status of the "missing" jobs that were moved. I managed to specify both servers with `qstat -f @awonmgr2 @flashmgr2` (the order matters for some reason, if I flip the orders of the servers it will only output the jobs on `awonmgr2`).

I can try to create an alias or a small "fake" `qstat` script that checks if it is executed with the `-f` flag and adds the 2 servers after the command.

Regarding your suggestions:
1. Can I change the code in my `conda`-installed `nextflow` or do I need to clone it from GitHub, change it and recompile it?
2. How can I provide those server arguments to `nextflow` to include them with `qstat`? They won't work with `qsub` (unless given as separate `-q` statements) so I can't just add them to `process.queue`.


Thanks, Ido

**Comment by pditommaso** [comment]
> 1. 

You need to clone, modify the source and compile it

> 2. 

In that case, the current implementation could be modified to split the `process.queue` by blanks and allowing to handle multiple queues. I think this could be a good solution.

**Comment by IdoBar** [comment]
Thanks @pditommaso,
It seems like `qstat` on the Awoonga HPC that I'm using is behaving very differently from other implementations that I've used: it only outputs just my jobs by default (like always using `-u $USER` argument) and it also has a bash wrapper to launch it that will handle the multiple servers. I couldn't find much documentation to that behaviour online, should ask my system admins.
For now I've edited the existing wrapper of `qstat` on the HPC to show the information for all servers (in the correct order), see this [gist](https://gist.github.com/IdoBar/f3aff3d82c7245bf8157f46153eb86e6).

A fairly simple modifications to the current code would be to parse the active server names and use that as arguments to `qstat -f`, something like this (adding the `-u $USER` flag will make this parse faster in busy systems):
```
qstat -u $USER -f $(qstat -B | egrep -v '(^Server|^---)' | awk -v ORS=" " '{print "@"$1}') 
``` 
I think that this would be the easiest solution at the moment, just need to double check that it works alright on all `PBSPro` systems.
I think that it would be difficult using the same `process.queue` parameter, which is mainly designed to control `qsub` behaviour (if multiple server/queue names are provided, which one is used with `qsub`?). An alternative is to add another `process.servers` or `process.qstat` parameter, that will allow to specify custom arguments for `qstat`.

Thanks, Ido
